### ISSUE TITLE
BundleDeployoments Status not being properly updated in Fleet local with Rancher 2.8-devel

### ISSUE THREAD

**Issue by mmartin24** [rancher/fleet#2128]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When deploying [fleet-examples ](https://github.com/rancher/fleet-examples) with path `simple` on local fleet cluster on Rancher 2.8-devel, it can take more than 20 mins to get clusters in Ready state. This does not seem only an UI issue as the backend displays the error on:

```
➜  ~ k get gitrepos -A -w
NAMESPACE     NAME               REPO                                        COMMIT                                     BUNDLEDEPLOYMENTS-READY   STATUS
fleet-local   time-health-test   https://github.com/rancher/fleet-examples   44f4634747e3dd6c9d1ad6c9d402b430f7aae20b   0/0                       
fleet-local   time-health-test   https://github.com/rancher/fleet-examples   44f4634747e3dd6c9d1ad6c9d402b430f7aae20b   0/0                       
fleet-local   time-health-test   https://github.com/rancher/fleet-examples   44f4634747e3dd6c9d1ad6c9d402b430f7aae20b   0/0                       WaitApplied(1) [Bundle time-health-test-simple]
fleet-local   time-health-test   https://github.com/rancher/fleet-examples   44f4634747e3dd6c9d1ad6c9d402b430f7aae20b   0/1                       WaitApplied(1) [Bundle time-health-test-simple]
fleet-local   time-health-test   https://github.com/rancher/fleet-examples   44f4634747e3dd6c9d1ad6c9d402b430f7aae20b   0/1                       NotReady(1) [Bundle time-health-test-simple]
fleet-local   time-health-test   https://github.com/rancher/fleet-examples   44f4634747e3dd6c9d1ad6c9d402b430f7aae20b   0/1                       NotReady(1) [Bundle time-health-test-simple]; deployment.apps default/frontend [progressing] Deployment does not have minimum availability., Available: 0/3; deployment.apps default/redis-master [progressing] Deployment does not have minimum availability., Available: 0/1; deployment.apps default/redis-slave [progressing] Deployment does not have minimum availability., Available: 0/2
fleet-local   time-health-test   https://github.com/rancher/fleet-examples   44f4634747e3dd6c9d1ad6c9d402b430f7aae20b   0/1                       NotReady(1) [Bundle time-health-test-simple]; deployment.apps default/frontend [progressing] Deployment does not have minimum availability., Available: 0/3; deployment.apps default/redis-master [progressing] Deployment does not have minimum availability., Available: 0/1; deployment.apps default/redis-slave [progressing] Deployment does not have minimum availability., Available: 0/2
fleet-local   time-health-test   https://github.com/rancher/fleet-examples   44f4634747e3dd6c9d1ad6c9d402b430f7aae20b   0/1                       NotReady(1) [Bundle time-health-test-simple]; deployment.apps default/frontend [progressing] Deployment does not have minimum availability., Available: 0/3; deployment.apps default/redis-master [progressing] Deployment does not have minimum availability., Available: 0/1; deployment.apps default/redis-slave [progressing] Deployment does not have minimum availability., Available: 0/2
fleet-local   time-health-test   https://github.com/rancher/fleet-examples   44f4634747e3dd6c9d1ad6c9d402b430f7aae20b   0/1                       NotReady(1) [Bundle time-health-test-simple]; deployment.apps default/frontend [progressing] Deployment does not have minimum availability., Available: 0/3; deployment.apps default/redis-master [progressing] Deployment does not have minimum availability., Available: 0/1; deployment.apps default/redis-slave [progressing] Deployment does not have minimum availability., Available: 0/2
fleet-local   time-health-test   https://github.com/rancher/fleet-examples   44f4634747e3dd6c9d1ad6c9d402b430f7aae20b   0/1                       NotReady(1) [Bundle time-health-test-simple]; deployment.apps default/frontend [progressing] Deployment does not have minimum availability., Available: 0/3; deployment.apps default/redis-master [progressing] Deployment does not have minimum availability., Available: 0/1; deployment.apps default/redis-slave [progressing] Deployment does not have minimum availability., Available: 0/2

```
![20_mins_not_ok](https://github.com/rancher/fleet/assets/37271841/6d11af37-6f57-4a81-b832-2c5412f3c4bc)
a

### Expected Behavior

Clusters should be ready in few minutes (<5 mins)

Workaround. Using "Force update" Fixes the issue as can be seen here:

https://github.com/rancher/fleet/assets/37271841/706891b9-3451-4dee-900b-e9783bc43567

### Steps To Reproduce

1. Install single node cluster with k3s `1.26.10+k3s2`
2. Install Rancher `v2.8-head`, I used:
```
helm repo add rancher-latest https://releases.rancher.com/server-charts/latest
helm repo update

helm upgrade --install rancher rancher-latest/rancher \
  --namespace cattle-system --create-namespace \
  --set global.cattle.psp.enabled=false \
  --set rancherImageTag=v2.8-head \
  --set hostname=$SYSTEM_DOMAIN \
  --set bootstrapPassword=password \
  --set replicas=1 \
  --wait
```
3. Go to `Continuous Delivery` > toggle to `fleet local ` > `Gitjob` > `Add Repository`
4. Add: 
    - Repository URL: https://github.com/rancher/fleet-examples
    - Branch name: `master`
    - Paths: Add Path > type `simple`
5. Click `Add`, then `Create`

### Environment

```markdown
- Architecture: amd
- Fleet Version: 
- Cluster:
  - Provider:k3s
  - Options: single cluster 
  - Kubernetes Version: `1.26.10+k3s2`
```

**Comment by manno** [comment]
This looks like it's related to the chart. Can you provide more debug information?
Maybe from the [fleet-agent logs](https://fleet.rancher.io/troubleshooting#fetch-the-log-from-the-fleet-agent), the [Gitrepo/Bundle status](https://fleet.rancher.io/troubleshooting#fetch-detailed-error-logs-from-gitrepos-and-bundles) and maybe the helm release on the downstream cluster, too?

**Comment by thehejik** [comment]
I consider this as a Critical Issue, it behaves the same on today's build of rancher `v2.8-882d1e32f9e5093f61aba4822fb2361dd61d66e6-head` with `fleet[-crd]:103.1.1+up0.9.1-rc.5` Application (pre)installed.

When adding PRIVATE HTTPS repo with BASIC AUTH for the first time, regardless if on local or downstream cluster, it is gonna fail and the status of the GitRepo is `Not Ready` but the resources are actually provisioned on the target cluster(s). 

The status can be hotfixed by selecting the GitRepo and pressing `Force Update`.

I could reproduce with simple nginx deployment stored on private Azure, Bitbucket and Gitlab repository. Surprisingly it works ok with private Github repo.

**Comment by thehejik** [comment]
Here you can see 15 Pods is up in `default` ns but fleet bundle reports `Available: 0/15`:
![image](https://github.com/rancher/fleet/assets/12828077/a3746bdd-3aaf-4462-bbdf-f883b03bff24)

**Comment by manno** [comment]
This only happens with basic auth credentials for the git repo?
And I guess, it does not affect Github repos, since they don't allow basic auth anymore?

**Comment by thehejik** [comment]
> This only happens with basic auth credentials for the git repo? And I guess, it does not affect Github repos, since they don't allow basic auth anymore?

Unfortunately I could reproduce it also with Private Azure DevOps SSH repo and keys on local `rancher:2.8.3-rc1` with fleet images:
```
rancher/fleet:v0.9.1-rc.5
rancher/gitjob:v0.9.3
rancher/fleet-agent:v0.9.1-rc.5
```
![image](https://github.com/rancher/fleet/assets/12828077/c4660af3-04d1-42da-83d1-b99f28868d8a)

```
$ kubectl get bundles -A
NAMESPACE     NAME                BUNDLEDEPLOYMENTS-READY   STATUS
fleet-local   fleet-agent-local   1/1                       
fleet-local   azure-private-ssh   0/1                       NotReady(1) [Cluster fleet-local/local]; deployment.apps default/fleet-nginx [progressing] Deployment does not have minimum availability., Available: 0/15
```
The pods are up and `Force Update` helped also this time.

**Comment by thehejik** [comment]
Private Github HTTPS repo is not affected.

**Comment by thehejik** [comment]
To me it looks like fleet-agent pod attempts to check fleet deployment from gitrepo and its pods availability 4 times and then it gives up just before the pods are scheduled.

```
cattle-fleet-local-system fleet-agent-75f9c4fd7d-flpd2 fleet-agent time="2024-02-28T13:02:59Z" level=error msg="bundle azure-private-https: deployment.apps default/fleet-nginx [progressing] Deployment does not have minimum availability., Available: 0/15"
```
Logs from all pods/jobs and namespaces [here](https://github.com/rancher/fleet/files/14434441/stern-fleet-azure-private.log)
Logs from all pods/jobs and namespaces after pressing `Force update` [here](https://github.com/rancher/fleet/files/14434448/stern-fleet-azure-private-force-update.log)
(produced by: `stern -A . --since 1s`)

**Comment by 0xavi0** [comment]
I've updated the subject of the issue as I could recreate using almost any `gitrepo`, no need to use private repos.

Investigating I realised the issue was because Modify events for `deployments` that were notifying for being ready were filtered, which made the `BundleDeployment` (and `Bundle`) to be reported as Not Ready. `gitrepos` take their status from `bundles` and that's why showed up as Not Ready.
At some other point (taking 30 minutes or so) there is a refresh in the cluster and that's when we finally see the real Ready status.
Issue was introduced in [this PR](https://github.com/rancher/fleet/pull/2031), which [was reverted](https://github.com/rancher/fleet/pull/2258) recently. 

It will be available when `Rancher 2.8.3` is released, so no stable release should be affected.

**Comment by 0xavi0** [comment]
# QA Template

## Solution
Solution was to rollback the PR that caused the issue

## Testing
Deploy any `gitrepo` with Fleet `v0.9.1-rc.2` to `v0.9.1-rc.6`.
For example:
```yaml
apiVersion: fleet.cattle.io/v1alpha1
kind: GitRepo
metadata:
  name: sample
  namespace: fleet-local
spec:
  repo: "https://github.com/0xavi0/fleet-examples"
  branch: master
  paths:
  - single-cluster/helm
```

The `gitrepo`, `bundle` and `bundleDeployment` don't reach the READY state although all deployments related (and pods) are ready.


Fix should be in future version `v0.9.1-rc.7` which should be released with `Rancher 2.8.3`.
With the fixed version everything should be back to normal. 


## Additional info

**Comment by mmartin24** [comment]
Fix noted.

Checked in  Rancher `2.8.3-rc6` with Fleet version `103.1.2+up0.9.2`
Tested using: `https://github.com/rancher/fleet-examples/tree/master/simple`
```yaml
kind: GitRepo
apiVersion: fleet.cattle.io/v1alpha1
metadata:
  name: simple
  namespace: fleet-local
spec:
  repo: https://github.com/rancher/fleet-examples
  paths:
  - simple
```

The example was deployed correctly showing 6 resources deployed in seconds....
![2024-03-27_15-44_1](https://github.com/rancher/fleet/assets/37271841/1b9fae12-2c6a-4b89-b447-1b1a2f53e378)



, as opposed as in `Rancher v2.8.0` with Fleet version `103.1.0+up0.9.0` where fails:
![2024-03-27_15-44_1](https://github.com/rancher/fleet/assets/37271841/8cad5124-58d5-4aa8-a458-96c3c2c0bb5f)

This test has been added to our UI e2e regression test Fleet-62 [here](https://github.com/rancher/fleet-e2e/pull/86).
